Better python binary detection

algernon opened the pull request:
Instead of using `AC_PATH_PROG` to find the python binary (which can only find binaries named `python`, not `python3` and similar), use `AM_PATH_PYTHON` instead, and take `--with-python` into account as well.

For this to work, we also need to call a few of our python scripts through the interpreter, to avoid having to generate the scripts.

Fixes #2090.

czanik commented:
Thanks! I am on mobile net right now but will check package building once back in my office.

kira-syslogng commented:
Build SUCCESS

czanik commented:
Thanks! **configure** ran fine and package is created for openSUSE Leap 15.0.

Next step (only tomorrow): testing the results.

Kokan commented on Mk/lex-rules.am:
```diff
@@ -1,6 +1,6 @@
 %.y: %.ym $(syslog_ng_tools)/merge-grammar.py $(syslog_ng_tools)/cfg-grammar.y
 	$(AM_V_at) $(mkinstalldirs) $(dir $@)
-	$(AM_V_GEN) $(syslog_ng_tools)/merge-grammar.py $< > $@
+	$(AM_V_GEN) $(PYTHON) $(syslog_ng_tools)/merge-grammar.py $< > $@
```
The CMake also has to use PYTHO for the grammar merge. Hint: `cmake/Modules/GenerateYFromYm.cmake`.

algernon replied:
Thanks, I missed that. Sorry!

algernon replied:
Fixed (I hope).

Kokan replied:
Maybe my comment was misleading the PYTHON variable in cmake does not exists, and does not point to the python interpreter. The `PYTHON_EXECUTABLE` does.

But still there is also a reason why cmake version won't work as intended.

The `with-python` and the option in cmake was not intended to specify which python version you are using for build, but rather for the module.

In cmake the python detection was "hidden" from the upper level, and determined in the python module. To follow up the same change in cmake as in autotools, one probably should move the python detection to the upper level in cmake.

Kokan replied:
Sure, I hope on next Monday I am going to be in the office, and we can check it together.

bazsi commented:
@kira-syslogng test this please

kira-syslogng commented:
Build SUCCESS

algernon commented:
Updated. Moved the CMake python detection out into the top-level `CMakeLists.txt`, so that `PYTHON_EXECUTABLE` will be available for the grammar generation stuff too.

kira-syslogng commented:
Build SUCCESS

Kokan commented on modules/python/CMakeLists.txt:
```diff
@@ -1,9 +1,4 @@
-set (PYTHON_VERSION "AUTO" CACHE STRING "Version of the installed development library" )
-
 if ("${PYTHON_VERSION}" STREQUAL "AUTO")
```
This condition is not needed anymore.

algernon commented:
Oof. I'll see about that, but that should be a separate PR, in my opinion. It "only" affects the tests, right?

Kokan commented:
No, it is all of the python module, well it says something like that, if the `PYTHON_VERSION` is auto, to whatever (do not check for python, but tries to compile).

algernon commented:
> No, it is all of the python module, well it says something like that, if the PYTHON_VERSION is auto, to whatever (do not check for python, but tries to compile).

Should have been clearer, my comment was a reply to [@mochrul's](https://github.com/balabit/syslog-ng/pull/2092#issuecomment-399756282).

kira-syslogng commented:
Build SUCCESS
